server/vulnerabilities/cpe.go

```diff
@@ -140,67 +115,391 @@ func cleanAppName(appName string) string {
+	}
+
+	return translations, nil
+}
+
+// DownloadCPETranslations downloads the CPE translations to the given vulnPath. If cpeTranslationsURL is empty, attempts to download it
+// from the latest release of github.com/fleetdm/nvd. Skips downloading if CPE translations is newer than the release.
+func DownloadCPETranslations(vulnPath string, client *http.Client, cpeTranslationsURL string) error {
+	path := filepath.Join(vulnPath, cpeTranslationsFilename)
+
+	if cpeTranslationsURL == "" {
+		release, err := GetLatestNVDRelease(client)
+		if err != nil {
+			return err
+		}
+		stat, err := os.Stat(path)
+		switch {
+		case errors.Is(err, os.ErrNotExist):
+			// okay
+		case err != nil:
+			return err
+		default:
+			if stat.ModTime().After(release.CreatedAt.Time) {
+				// file is newer than release, do nothing
+				return nil
+			}
+		}
+
+		for _, asset := range release.Assets {
+			if cpeTranslationsFilename == asset.GetName() {
+				cpeTranslationsURL = asset.GetBrowserDownloadURL()
+				break
+			}
+		}
+		if cpeTranslationsURL == "" {
+			return errors.New("failed to find cpe translations in nvd release")
+
+		}
+	}
+
+	u, err := url.Parse(cpeTranslationsURL)
+	if err != nil {
+		return err
+	}
+	if err := download.Download(client, u, path); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// regexpCache caches compiled regular expressions. Not safe for concurrent use.
+type regexpCache struct {
+	re map[string]*regexp.Regexp
+}
+
+func newRegexpCache() *regexpCache {
+	return &regexpCache{re: make(map[string]*regexp.Regexp)}
+}
+
+func (r *regexpCache) Get(pattern string) (*regexp.Regexp, error) {
+	if re, ok := r.re[pattern]; ok {
+		return re, nil
+	}
+
+	re, err := regexp.Compile(pattern)
+	if err != nil {
+		return nil, err
+	}
+	r.re[pattern] = re
+	return re, nil
+}
+
+// CPETranslations include special case translations for software that fail to match entries in the NVD CPE Dictionary
+// using the standard logic. This may be due to unexpected vendor or product names.
+//
+// Example:
+//
+//     [
+//       {
+//         "match": {
+//           "bundle_identifier": ["com.1password.1password"]
+//         },
+//         "translation": {
+//           "product": ["1password"],
+//           "vendor": ["agilebits"]
+//         }
+//       }
+//     ]
+type CPETranslations []CPETranslationEntry
+
+func (c CPETranslations) Translate(reCache *regexpCache, s *fleet.Software) (CPETranslation, bool, error) {
+	for _, entry := range c {
+		match, err := entry.Match.Matches(reCache, s)
+		if err != nil {
+			return CPETranslation{}, false, err
+		}
+		if match {
+			return entry.Translation, true, nil
+		}
+	}
+
+	return CPETranslation{}, false, nil
+}
+
+type CPETranslationEntry struct {
+	Match       CPETranslationMatch `json:"match"`
+	Translation CPETranslation      `json:"translation"`
+}
+
+// CPETranslationMatch represents match criteria for cpe translations.
+type CPETranslationMatch struct {
+	Name             []string `json:"name"`
```

Comment:
Let's document `Name` is a regex.

Reply:
Hmm, do you think we should put it in `server/vulnerabilities/README.md` or somewhere in `docs/`? We have some documentation on vulnerability processing in `docs/Using-Fleet/Vulnerability-Processing.md`, so maybe that would be a good place.

Reply:
Latter sounds good.

Reply:
Updated docs. I also ended up renaming `translation` to `filter` because the name was already overused. Noticed this when I was trying to write some docs.